akka-projection-core/src/main/scala/akka/projection/Projection.scala

```diff
@@ -66,18 +66,34 @@ private[projection] object RunningProjection {
         () => source()
       }
 
-  def stopHandlerWhenStreamCompletedNormally(whenStreamCompleted: Future[Done], stopHandler: () => Future[Done])(
-      implicit ec: ExecutionContext): Future[Done] = {
-    whenStreamCompleted.flatMap(_ => stopHandler())
-  }
+  /* internal exception to wrap exceptions coming from stopHandler */
+  private case class StopHandlerException(cause: Throwable) extends RuntimeException(cause) with NoStackTrace
+
+  /**
+   * Adds a `watchTermination` on that passed Source that will call the `stopHandler` on completion.
+   *
+   * The stopHandler function is called on success or failure. In case of failure, the original failure is preserved.
+   */
+  def stopHandlerOnTermination(src: Source[Done, NotUsed], stopHandler: () => Future[Done])(
+      implicit ec: ExecutionContext): Source[Done, Future[Done]] = {
+    src.watchTermination() { (_, futDone) =>
+      futDone
+        .flatMap { _ =>
+          stopHandler().recoverWith {
+            // if stop fails we need to wrap it so
+            // on the next recoverWith we don't call it twice
+            case exc => Future.failed(StopHandlerException(exc))
+          }
+        }
+        .recoverWith {
+          case StopHandlerException(exc) => Future.failed(exc)
+          case streamFailure             =>
+            // ignore error in stop failure and preserve original stream failure
+            stopHandler().recoverWith(_ => Future.failed(streamFailure))
+        }
```

Comment:
On my first impl, I was first recovering and then flatMapping. 

For a test where we forced the failing, the stop method was being called twice. You can't imagine the amount of println and wasted time here. 

I hope the code is making it clear why that dance is need. If someone has better ideas, let me know. I find it a little verbose.